### ISSUE TITLE
exportRecords() recasts checkbox labels as numeric even when not numeric

### ISSUE THREAD

**Issue by tobadia** [nutterb/redcapAPI#92]
Hi again @nutterb 

I have recently come across a bug that silently drops in `exportRecords()` when `factors=FALSE`, or otherwise fails with an error when `factors=TRUE`.
In the code for the `exportRecords()`routine, the following is applied to variables that are of class checkbox:
```
    #* This function grabs the variable codings of the checkbox variables.
    #* These need to be appended to the field names
    #* In this process, a checkbox field name is replicated as many times as it has options
    checkvars <- function(x){
      if (meta_data$field_type[meta_data$field_name %in% x] == "checkbox"){
        opts <- unlist(strsplit(meta_data$select_choices_or_calculations[meta_data$field_name %in% x], "[|]"))
        opts <- tryCatch(as.numeric(unlist(sapply(strsplit(opts, ","), '[', 1))),
                         warning = function(cond){ nm <- as.character(unlist(sapply(strsplit(opts, ","), '[', 1)))
                                                   nm <- gsub('^\\s*','',nm,perl=TRUE)
                                                   nm <- gsub('\\s*$','',nm,perl=TRUE)
                                                   return(nm)})
        # negative integers are permitted for checkbox values, 
        # but the API converts the - to a _
        x <- paste(x, sub("[-]", "_", opts), sep="___")
      }
      return(x)
    }
    field_names <- sapply(field_names, checkvars)
```
In te tryCatch function, the fariable names are assumed to be stricly of numeric type. However, REDCap also allows for character strings :
```
You may manually define the coded value for each choice by entering the coded number and a comma before the choice label (see example below). The only acceptable raw values are 1) an integer, or 2) a text string containing only letters, numbers, and underscores.
```

In case where characters are used, or numeric with leading zeros for consistency in number of characters, the checkvars() function will return non-existent field_names that will subsequently be passed to the rest of the code. Worst case scenario is when `factors=FALSE`, the content of the columns of the data.frame is silently result to an empty vector.

I have prepared a small PR that seems to work fine with usual integers, while also allowing characters and leading zeros (or whatever that complies with REDCap coding of checkboxes).

Hope that helps!

Thomas

**Comment by nutterb** [comment]
I can see where this was an issue in the master branch.  There are a lot of difficult-to-merge changes in the Version2 branch.  Could you test this after running the code below and let me know if the issue still exists?

(I, unfortunately, don't really have a REDCap installation I can use for thorough testing anymore)

`devtools::install_github("nutterb/redcapAPI", ref = "Version2")`

**Comment by tobadia** [comment]
It seems to work fine when `factors = FALSE` however yields something strange (to me) when `factors=TRUE`, though I could be mistaken. What I would be expecting, based on man pages:

- `factors = FALSE, checkboxLabels = FALSE` : working as intended, values are 0 or 1 depending on if item is checked or unchecked
- `factors = TRUE, checkboxLabels = FALSE` : not working as intended, values should be "Checked" / "Unchecked" depending if item is checked or not, but they are actually "_value_" and "" (where _value_ is the coding value of the checkbox element as defined by REDCap)
- `factors = TRUE, checkboxLabels = TRUE` : working as intended, it returns "" or "_label_" this time, with label the item displayed as checkbox label

Basically it seems to work OK except in the default case for `checkboxLabels`, where blanks are returned instead of "Unchecked". Maybe for consistency the output should be similar regardless of `checkboxLabels` ? _i.e._ either blank/not-blank or not-blank/not-blank, but I understand it yields an issue as to what label you'd assign to non-checked boxes then. In any case, that's a secondary reflexion.

Thanks for looking into it. If that can help, we have a server in my institution that I can use to test some stuffs, do not hesitate to let me know if you want me to test features you're not sure are running fine!

**Comment by nutterb** [comment]
You're right, something in there seems to be a little bit off.  Having read the documentation again now (perhaps for the first time since I wrote it), I'm thinking it probably ought to look more like the behavior in the right column.  Does that seem reasonable to you?

| `factors` | `checkboxLabels` | Documented Behavior | Observed Behavior | Ideal Behavior |
|----------|--------------------|--------------------------|----------------------|-----------------|
| FALSE    | FALSE                   | 0/1                               | 0/1                          | 0/1                   |
| FALSE    | TRUE                    | "" / label (character)     | N/A                        | "" / value (character) |
| TRUE     | FALSE                   | Checked/Unchecked    | "" / value                | Checked / Unchecked |
| TRUE     | TRUE                    | "" / label (factor)          | "" / label                 | "" / label (factor) |

I agree that the blank/not-blank format is a little weird.  To be fair to myself, it isn't something I came up with.  It was intended to mimic the `checkboxLabels` option of the API export from REDCap itself. I've never been fully satisfied with how the checkbox output is displayed, but I also recognize that the display I want is largely context dependent.  Hence the `recodeCheck` function, which let's me change the levels at will with relative easy.

**Comment by tobadia** [comment]
This does make sense indeed. As you point out, the blank/non-blank output
is a little odd but it is definitely a good idea to have an output that
mimics the behaviour of the API. In all fairness, I haven't been through
the whole API documentation, but the more consistent the outputs, the
better.

The most common use casse would be rows 1 and 3 in your table just above
anyway, which make more sense in terms of data structure and "readability".

I very much apreciate the time you share for these small fixes and
understand that it's not easy maintaining the package if you don't have an
easy access to a redcap server these days. So just in case, apologies if
I'm being picky on this stuff, and thanks for the open discussion :-)

2018-01-02 21:10 GMT+01:00 Benjamin <notifications@github.com>:

**Comment by nutterb** [comment]
Reopening just so I don't forget to work this in on Friday.

**Comment by nutterb** [comment]
I'm 99% sure I've got this, but wouldn't mind some independent validation. 

Thanks.

**Comment by tobadia** [comment]
Will do over the weekend, left early due to holiday. But since I gotta
monitor our ongoing studies, laptop is in the trunk. If I get the chance
I'll do it today but not really sure.
Thanks again

Le 5 janv. 2018 13:56, "Benjamin" <notifications@github.com> a écrit :

I'm 99% sure I've got this, but wouldn't mind some independent validation.

Thanks.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/nutterb/redcapAPI/issues/92#issuecomment-355548573>,
or mute
the thread

**Comment by nutterb** [comment]
Do it next week.   Use your time off for time off.  I have until the 21st to submit, so it isn't that much of a rush.

**Comment by tobadia** [comment]
Works as described in previous post, thanks!
